src/main/resources/docker-entrypoint.sh

```diff
@@ -0,0 +1,206 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Requirements:
+# 1) Exec as user
+# * Under Alpine: su-exec (e.g. apk add --no-cache su-exec)
+# * Otherwise: gosu (e.g. apt-get install gosu)
+# 2) Add user and group
+# * Under Alpine: adduser, addgroup
+# * Otherwise: useradd, groupadd
+
+# Exit on any command error
+set -e
+
+# Whether to log during execution
+verbose="true"
+
+# The command and arguments to execute
+command_args=
+
+# The configuration for docker entrypoint
+dep_configuration=
+
+log_err() {
+  l_prefix=$(date  +'%H:%M:%S')
+  printf "[%s] %s\\n" "${l_prefix}" "$@" 1>&2;
+}
+
+log_out() {
+  if [ -n "${verbose}" ]; then
+    l_prefix=$(date  +'%H:%M:%S')
+    printf "[%s] %s\\n" "${l_prefix}" "$@"
+  fi
+}
+
+assert_user_group_matches() {
+  l_name="$1"
+  l_uid="$2"
+  l_gid="$3"
+
+  l_existing_uid=$(id -u "${l_name}")
+  if [ "${l_existing_uid}" != "${l_uid}" ]; then
+    # The user exists but the user id does not match
+    log_err "The user ${l_name} already exists but with uid ${l_existing_uid} and not ${l_uid}"
+    exit 1
+  fi
+
+  l_existing_gid=$(id -g "${l_name}")
+  if [ "${l_existing_gid}" != "${l_gid}" ]; then
+    # The user exists but the group id does not match
+    log_err "The user ${l_name} already exists but with gid ${l_existing_gid} and not ${l_gid}"
+    exit 1
+  fi
+}
+
+linux_create_user_group() {
+  l_name="$1"
+  l_uid="$2"
+  l_gid="$3"
+
+  if id -u "${l_name}" > /dev/null 2>&1; then
+    # Check that the existing user's uid and gid match the given ones
+    assert_user_group_matches "${l_name}" "${l_uid}" "${l_gid}"
+
+    # The user exists with matching user and group id
+    log_out "The user ${l_name} already exists with required ids ${l_existing_uid}:${l_existing_gid}"
+    return
+  fi
+
+  log_out "Creating user ${l_name} (${l_uid}) as member of group ${l_name} (${l_gid})"
+  groupadd -r "${l_name}" "--gid=${l_gid}"
+  useradd -r -g "${l_name}" "--uid=${l_uid}" "${l_name}"
+}
+
+alpine_create_user_group() {
+  l_name="$1"
+  l_uid="$2"
+  l_gid="$3"
+
+  if id -u "${l_name}" > /dev/null 2>&1; then
+    # Check that the existing user's uid and gid match the given ones
+    assert_user_group_matches "${l_name}" "${l_uid}" "${l_gid}"
+
+    # The user exists with matching user and group id
+    log_out "The user ${l_name} already exists with required ids ${l_existing_uid}:${l_existing_gid}"
+    return
+  fi
+
+  log_out "Creating user ${l_name} (${l_uid}) as member of group ${l_name} (${l_gid})"
+  addgroup -g "${l_gid}" -S "${l_name}"
```

Comment:
useradd should be on both

Reply:
Are you sure?

```
addgroup [-g GID] [-S] [USER] GROUP
```

https://wiki.alpinelinux.org/wiki/Setting_up_a_new_user

```
$ docker run -it alpine:latest /bin/sh
Unable to find image 'alpine:latest' locally
latest: Pulling from library/alpine
df20fa9351a1: Pull complete 
Digest: sha256:185518070891758909c9f839cf4ca393ee977ac378609f700f60a771a2dfe321
Status: Downloaded newer image for alpine:latest
/ # addgroup -g 1234 -S test
/ # adduser -u 2345 -S -G test test
/ # grep test /etc/passwd
test:x:2345:1234:Linux User,,,:/home/test:/sbin/nologin
/ # grep test /etc/group
test:x:1234:test
```

Reply:
I didn't say adduser wasn't there. Just that useradd should be on both so you can remove differentiation.

Reply:
Oh ;)

The image used for kdb (non-alpine) has both:

```
root@1ac35a78c373:/opt/kairosdb# which useradd
/usr/sbin/useradd
root@1ac35a78c373:/opt/kairosdb# which adduser
/usr/sbin/adduser
root@1ac35a78c373:/opt/kairosdb# cat /etc/issue
Debian GNU/Linux 10 \n \l
```

Alpine only has `adduser`:

```
/ # which adduser
/usr/sbin/adduser
/ # which useradd
/ # echo $?
1
```

Here is the help on Alpine:

```
/ # adduser --help
BusyBox v1.31.1 () multi-call binary.

Usage: adduser [OPTIONS] USER [GROUP]

Create new user, or add USER to GROUP

	-h DIR		Home directory
	-g GECOS	GECOS field
	-s SHELL	Login shell
	-G GRP		Group
	-S		Create a system user
	-D		Don't assign a password
	-H		Don't create home directory
	-u UID		User id
	-k SKEL		Skeleton directory (/etc/skel)
```

And here it is on the kdb base image (debian I think):

```
adduser --help
adduser [--home DIR] [--shell SHELL] [--no-create-home] [--uid ID]
[--firstuid ID] [--lastuid ID] [--gecos GECOS] [--ingroup GROUP | --gid ID]
[--disabled-password] [--disabled-login] [--add_extra_groups] USER
   Add a normal user

adduser --system [--home DIR] [--shell SHELL] [--no-create-home] [--uid ID]
[--gecos GECOS] [--group | --ingroup GROUP | --gid ID] [--disabled-password]
[--disabled-login] [--add_extra_groups] USER
   Add a system user

adduser --group [--gid ID] GROUP
addgroup [--gid ID] GROUP
   Add a user group

addgroup --system [--gid ID] GROUP
   Add a system group

adduser USER GROUP
   Add an existing user to an existing group

general options:
  --quiet | -q      don't give process information to stdout
  --force-badname   allow usernames which do not match the
                    NAME_REGEX configuration variable
  --help | -h       usage message
  --version | -v    version number and copyright
  --conf | -c FILE  use FILE as configuration file
```

The alpine versions of these appear different. The same goes for `addgroup`.